tasktiger/worker.py

```diff
@@ -569,12 +575,34 @@ def check_child_exit():
 
                 # Wait until the timeout or a signal / child exit occurs.
                 try:
-                    select.select(
+                    # If observed the following behavior will be seen
+                    # in the pipe when the parent process receives a
+                    # SIGTERM while a task is running in a child process:
+                    # Linux:
+                    #   - 0 when parent receives SIGTERM
+                    #   - select() exits with EINTR when child exit
+                    #     triggers signal, so the signal in the
+                    #     pipe is never seen since check_child_exit()
+                    #     will see the child is gone
+                    #
+                    # macOS:
+                    #   - 15 (SIGTERM) when parent receives SIGTERM
+                    #   - 20 (SIGCHLD) when child exits
+                    results = select.select(
                         [pipe_r],
                         [],
                         [],
                         self.config['ACTIVE_TASK_UPDATE_TIMER'],
                     )
+
+                    if results[0]:
+                        # Purge pipe so select will pause on next call
+                        try:
+                            opened_fd.read(1)
```

Comment:
I tried and it would just return an empty string, rather than raise an `IOError`. Obviously the code is still going to work as is even if we catch the error.

```
In [1]: import os                                                                                                                                                                                                 
                                                                                                                                                                                                                  
In [2]: pipe_r, pipe_w = os.pipe()

In [3]: opened_fd = os.fdopen(pipe_r)

In [5]: import fcntl

In [6]: flags = fcntl.fcntl(pipe_r, fcntl.F_GETFL, 0)
   ...: 

In [7]: flags = flags | os.O_NONBLOCK
   ...: 

In [8]: fcntl.fcntl(pipe_r, fcntl.F_SETFL, flags)
Out[8]: 0

In [9]: opened_fd.read(1)
Out[9]: ''

In [10]: opened_fd.read(1)
Out[10]: ''
```

Reply:
Linux raises the exception, Mac doesn't.  I added a note since I explain the signal behavior above for both platforms and this is relevant also.

```
In [1]: import os

In [2]: pipe_r, pipe_w = os.pipe()

In [3]: opened_fd = os.fdopen(pipe_r)

In [4]: import fcntl

In [5]: flags = fcntl.fcntl(pipe_r, fcntl.F_GETFL, 0)

In [6]: flags = flags | os.O_NONBLOCK

In [7]: fcntl.fcntl(pipe_r, fcntl.F_SETFL, flags)
Out[7]: 0

In [8]: opened_fd.read(1)
---------------------------------------------------------------------------
IOError                                   Traceback (most recent call last)
/home/closeio/closeio/closeio/exporter/documents/__init__.py in <module>()
----> 1 opened_fd.read(1)

IOError: [Errno 11] Resource temporarily unavailable

In [9]:
```

Reply:
Turns out Python 2.7 and 3.x behavior is different in Linux. 3.x doesn't raise an exception probably due to [PyErr_Clear()](https://github.com/python/cpython/blob/3.8/Modules/_io/fileio.c#L815) in Py3 that doesn't exist in the [2.7 implementation](https://github.com/python/cpython/blob/2.7/Modules/_io/fileio.c#L694-L695)?